src/main/java/com/commercetools/sync/services/impl/BaseService.java

```diff
@@ -135,4 +153,156 @@
+     * @param typeName  a resource type name, f.e.: <i>ProductType</i>
+     * @return {@link CompletionStage}&lt;{@link Optional}&gt; in which the result of it's completion contains an
+     *         {@link Optional} that contains the matching id if exists, otherwise empty.
+     */
+    CompletionStage<Optional<String>> fetchAndCache(@Nullable final String key,
+                                                    @Nonnull final Supplier<Q> query,
+                                                    @Nonnull final Function<U, String> keyMapper,
+                                                    @Nonnull final String typeName) {
+        final Consumer<List<U>> resourcePageConsumer = resourcePage ->
+            resourcePage.forEach(resource -> {
+                final String fetchedResourceKey = keyMapper.apply(resource);
+                final String id = resource.getId();
+                if (StringUtils.isNotBlank(fetchedResourceKey)) {
+                    keyToIdCache.put(fetchedResourceKey, id);
+                } else {
+                    syncOptions.applyWarningCallback(format("%s with id: '%s' has no key set. Keys are"
+                        + " required for resource matching.", typeName, id));
+                }
+            });
+
+        return CtpQueryUtils
+            .queryAll(syncOptions.getCtpClient(), query.get(), resourcePageConsumer)
+            .thenApply(result -> Optional.ofNullable(keyToIdCache.get(key)));
+    }
+
+    /**
+     * Given a {@link Set} of resource keys, this method fetches a set of all the resources, matching this given set of
+     * keys in the CTP project, defined in an injected {@link SphereClient}. A mapping of the key to the id
+     * of the fetched resources is persisted in an in-memory map.
+     *
+     * @param resourceKeys set of state keys to fetch matching states by
+     * @param query        a function to get query
+     * @param keyMapper    a function to get the key from the returned resource
+     * @return {@link CompletionStage}&lt;{@link Set}&lt;{@code U}&gt;&gt; in which the result of it's completion
+     *         contains a {@link Set} of all matching resources.
+     */
+    @Nonnull
+    CompletionStage<Set<U>> fetchMatchingResources(@Nonnull final Set<String> resourceKeys,
+                                                   @Nonnull final Supplier<Q> query,
+                                                   @Nonnull final Function<U, String> keyMapper) {
+        if (resourceKeys.isEmpty()) {
+            return CompletableFuture.completedFuture(Collections.emptySet());
+        }
+
+        return CtpQueryUtils
+            .queryAll(syncOptions.getCtpClient(), query.get(), Function.identity())
+            .thenApply(fetchedResources -> fetchedResources
+                .stream()
+                .flatMap(List::stream)
+                .peek(resource -> keyToIdCache.put(keyMapper.apply(resource), resource.getId()))
```

Comment:
Ideally it would be best to do: 
````java
keyToIdCache.put(resource.getKey(), resource.getId());
````
instead of having to pass a `keyMapper`. 

but since the `resource` is of type `Resource` which doesn't have the `getKey()` method. We cannot do that.
However, what we could do is: have the generic type `U` not only extend Resource but also the `WithKey` interface which gives us accessibility to the `getKey` method.
```java
U extends Resource<U> & WithKey
```

This will give us the getKey method 😉**BUT** there is a problem with that from the JVM SDK itself is that the `Category` and `Product` resources don't extend this interface, but they should. I will make an issue out of it on the JVM SDK. Until then, I guess we can use the mapper here :)

Reply:
Issue opened here: https://github.com/commercetools/commercetools-jvm-sdk/issues/1957

Reply:
At first I wanted to use `WithKey`, but like you said not all resources are extend it that's why keyMapper was needed.

Reply:
it would be nice to add a TODO comment to linking this github issue so that we can address it later :)

Reply:
Fixed: [e4c891d](https://github.com/commercetools/commercetools-sync-java/pull/419/commits/e4c891d1d4a14e94f7bf9deb70b3ad64e2da639f)